### DISCUSSION TITLE
Vercel Dev command in monorepo - how to use?

### DISCUSSION THREAD

**Discussion by dszymczuk** [vercel/vercel#5294] [Help] [answered]
Hi,

I have problem with `vercel dev` in my project.
My structure of repo:
```
ROOT
|- CLIENT (create-react-app)
|- API (empty for now :) )
|- PLUGINS (serverless)

```
I've created two projects with the same repo because I would like to have the **client** on different link and **plugins** on different. I called the projects as **PROJECT-CLIENT** and **PROJECT-PLUGINS**.
For **PROJECT-CLIENT** I've set root directory to **CLIENT**.
For **PROJECT-PLUGINS** I've set root directory to **PLUGINS**.
From GitHub perspective, everything works fine.

I've linked by `vercel` command ROOT folder to **PROJECT-CLIENT** and PLUGINS folder to **PROJECT-PLUGINS**.
I can deploy project independent (run vercel command from **ROOT** and **PLUGINS**).
But I would like to run `vercel dev` in **PLUGINS** to handle serverless functions.
When I run `vercel dev` in PLUGINS folder, I've got `Error! /PROJECT_PATH/plugins/plugins doesn't exist`.
So I've tried from ROOT folder and `vercel dev ` started from **CLIENT** path.
When I run with path `vercel dev PLUGINS`, I've got `Error! /PROJECT_PATH/plugins/plugins doesn't exist`.

I can't remove **PLUGINS** folder from Root Directory because it will destroy auto deployment from GitHub.

How can I achieve all of these:
auto deployment from GitHub **CLIENT** folder -> **PROJECT-CLIENT**
auto deployment from GitHub **PLUGINS** folder -> **PROJECT-PLUGINS**
`vercel dev` run for PLUGINS

**Comment by julioflima** [comment]
Same problem here.

In folder CLIENT: `nano .vercel`. Resulted a blank file.

**Comment by styfle** [comment, marked as answer]
At this time, `vercel dev` only works from the **root of the repository** and only one project at a time.

Run `vercel link` to select your project, then `vercel dev` to develop (both from the root of the repo).

**Reply by Beraliv** [reply]
@styfle  Thank you, helped me!

**Comment by joaovpmamede** [comment]
Do you know if has anything changed since your last comment @styfle?

**Reply by styfle** [reply]
Not yet, but @Ethan-Arrowood is currently working on improving monorepo support.

**Reply by RyadMM** [reply]
> Not yet, but @Ethan-Arrowood is currently working on improving monorepo support.

Hi @styfle @Ethan-Arrowood  Curious if you have an update on this?

**Reply by abhishiv** [reply]
Does anyone have a workaround in the meantime? I'm stuck at deleting the .vercel folder and running one pkg one at a time so my workflow really sucks.

**Reply by RyadMM** [reply]
> Does anyone have a workaround in the meantime? I'm stuck at deleting the .vercel folder and running one pkg one at a time so my workflow really sucks.

We run each vercel serverless api inside their own container.

Note that this is exclusively for dev, and not for production. That is why it is not really optimized. However I am open to feedback and suggestions!

So for example dockerfile for app A:

```dockerfile
FROM node:16-alpine

RUN apk add --no-cache git make clang build-base python3

WORKDIR /app/

COPY .env.dev /app/.env
COPY package.json yarn.lock /app/
COPY ./apps/app-a/package.json /app/apps/app-a/package.json
COPY .yarnrc.yml /app/.yarnrc.yml
COPY .yarn /app/.yarn
COPY ./apps/app-a/scripts/start.sh /start.sh

RUN yarn workspace app-a install

COPY ./apps/app-a/ /app/apps/app-a/

EXPOSE 3001

ENTRYPOINT ["/bin/sh", "/start.sh"]
```

the start.sh script:
```bash
#!/bin/sh
yarn vercel link /app/apps/app-a/ -p app-a --token ${VERCEL_TOKEN} --scope ${SCOPE} --yes
mv /app/apps/app-a/.vercel /app/.vercel
yarn vercel dev --listen 3001 --token ${VERCEL_TOKEN} --yes
```

In my team we have three vercel APIs that follow this format, each one referenced inside a docker-compose.yml.

Then docker-compose up run all apps.



It might be a bit complicated but fits our use case since we were already containerizing the APIs.

**Reply by emirotin** [reply]
The current state is quite unfortunate.

By the way things are advertised I had the expectations that Vercel cloud + Vercel CLI + Turborepo + pnpm + Next.js/Serverless functions (with pnpm being supported by Vercel and the rest owned by it) would be a seamless experience.

I already have 1 Next.js and 1 pure API project in my monorepo, with more API projects expected to be added soon, and now I'm unable to work on several of them in parallel (the greatest selling point of a monorepo) without hacking around with the dockerfiles and docker-compose :/

**Reply by abhishiv** [reply]
@emirotin Nice to see a fellow talliner! In my case, I ended up moving to cloudflare pages + workers due to this issue since wrangler has good support for this. Maybe that could be an option for you?

**Reply by emirotin** [reply]
@abhishiv thanks for the tip!

In the end, I have figured the alpha-level yet working feature of the CLI, `link --repo`. It lets you link a monorepo to multiple vercel projects and after that things seem to work.
